---
title: "Introduction to M3Drop: Michaelis-Menten modelling of dropouts in scRNASeq"
author: "Tallulah Andrews"
date: 13 July 2016
output:
    BiocStyle::html_document
---

```{r knitr-options, echo=FALSE, warning=FALSE}
## To render an HTML version that works nicely with github and web pages, do:
## rmarkdown::render("vignettes/vignette.Rmd", "all")
library(knitr)
opts_chunk$set(fig.align = 'center', fig.width = 6, fig.height = 5, dev = 'png')
#knitr::opts_chunk$set(echo=FALSE, fig.path='cellity/plot-', cache=TRUE)
theme_set(theme_bw(12))
```

# Introduction
Single-cell RNA sequencing is able to quantify the whole transcriptome from the small amount of RNA present in individual cells. However, a consequence of reverse-transcribing and amplifying small quantities of RNA is a large number of dropouts, genes with zero expression in particular cells. The frequency of dropout events is strongly non-linearly related to the measured expression levels of the respective genes. M3Drop posits that these dropouts are due to failures of reverse transcription, a simple enzyme reaction, thus should be modelled using the Michaelis-Menten equation as follows:

$$P_{i} = 1 - \frac{S_{i}}{S_{i}+K}$$

Where $P_{i}$ is the proportion of cells where gene $i$ dropouts out, $S_{i}$ is the mean expression of gene $i$ and $K$ is the Michaelis constant.

# Installation
M3Drop can be installed directly from Github.
```{r eval=FALSE}
library(devtools)
install_github('tallulandrews/M3D')
```
or from Bioconductor
```{r eval=FALSE}
#Coming Soon!
#source("http://bioconductor.org/biocLite.R")
#biocLite("M3Drop")
```
or from the package source
```{r eval=FALSE}
install.package('M3Drop_0.99.0.tar.gz', type="source")
```

# Example Workflow
We'll be using a portion of the Deng et al. (2014) dataset in this example. You can download the R-package containing this data [here](https://github.com/tallulandrews/M3DExampleData.git).
```{r eval=TRUE,results=FALSE}
library(M3Drop)
library(M3DExampleData)
```

## QC and Normalization
The first step is to clean the data by remove cells with too few detected genes, genes that with very low expression, and to normalize the data. This can be done using any method but M3Drop includes a simple function that will clean the expression matrix and convert raw counts to counts per million (CPM).

```{r eval=TRUE} 
# Remove cells with <2,000 detected genes and apply CPM normalization
Normalized_data = M3Drop_Clean_Data(Mmus_example_list$data, labels = Mmus_example_list$labels, 
					is.counts=TRUE, min_detected_genes=2000)
dim(Normalized_data$data)
length(Normalized_data$labels)
```

## Fitting the Michaelis-Menten
Next, we can compare the fits of different possible functions relating the proportion of dropouts to the average expression for each gene.

```{r}
fits = M3Drop_Dropout_Models(Normalized_data$data)
```
**Figure 1**: Fits of three different dropout models.

Visual inspection of the resulting plot shows that the Michaelis-Menten equation is the best fit to the data. However, we can also examine some statistics to confirm this:
```{r}
# Sum of absolute residuals
data.frame(MM=fits$MMFit$SAr,Logistic=fits$LogiFit$SAr,DoubleExpo=fits$ExpoFit$SAr)
# Sum of squared residuals
data.frame(MM=fits$MMFit$SSr,Logistic=fits$LogiFit$SSr,DoubleExpo=fits$ExpoFit$SSr)
```
Here we see that the sum of squared residuals favours the flatter logistic curve due to the noise in the data, where as sum of absolute residuals shows the Michaelis-Menten is the best fit to the data.

## Identifying Differentially Expressed (DE) Genes

Since the Michaelis-Menten equation is concave, averaging across a mixed population forces differentially expressed genes to be shifted to the right of the Michaelis-Menten curve. DE genes are identified by comparing the local $K$ calculated for a specific gene to the global $K$ fitted across all genes using a Z-test followed by multiple-testing correction. Here we find 1,248 DE genes at 1% FDR.

```{r}
DE_genes = M3Drop_Differential_Expression(Normalized_data$data, 
			mt_method="fdr", mt_threshold=0.01)
```
**Figure 2**: Differentially expressed genes at 1% FDR (purple).

Note that this function runs directly from the expression matrix, hence one could skip straight to identifying DE genes without comparing models and any external normalisation method can be applied to the raw counts prior to DE analysis. 

## Examining DE Genes and Identifying Subpopulations of Cells

To check that the identified genes are truly differentially expressed we can plot the normalised expression of the genes across cells.

```{r}
heat_out = M3Drop_Expression_Heatmap(DE_genes$Gene, Normalized_data$data, 
			cell_labels = Normalized_data$labels)
```
**Figure 3**: Heatmap of expression of DE genes.

The heatmap shows that the identified DE genes are differentially expressed across timepoints. Furthermore, it shows that the blastocysts cluster into two different groups based on the expression of these genes. We can extract these groups and identify marker genes for them as follows:

```{r}
cell_populations = M3Drop_Get_Heatmap_Cell_Clusters(heat_out, k=4)
library("ROCR") # Calculating AUC of markers requires ROCR package
marker_genes = M3Drop_getmarkers(Normalized_data$data, cell_populations)
```

The first function cuts the dendrogram from the heatmap to produce k clusters of cells. These labels are stored in cell_populations. The second function tests all genes as marker genes for the provided clusters.

Marker genes are ranked by the area-under the ROC curve (AUC) for predicting the population with the highest expression of the gene from the other groups. Significance is calculated using a Wilcox-rank-sum test. Now we can examine the marker genes of the two clusters of blastocyst cells more closely.

```{r}
head(marker_genes[marker_genes$Group==3,],20) # Top twenty marker genes
marker_genes[rownames(marker_genes)=="Cdx2",] # Get marker data for Cdx2
```

This shows that the inner cell mass (ICM) marker Sox2 is one of the top 20 markers for group 3 and that the trophectoderm (TE) marker Cdx2 is a marker for group 2, suggesting these two clusters coorespond to ICM and TE cells within each blastocyst.

## Comparing to Other Methods

For comparison purposes, I have also included a function which implements the method to identify highly variable genes presented in Brennecke et al. (2013) with the added option to run it without providing spike-ins, in which case all genes are used to fit the function between CV2 and mean expression.

```{r}
HVG = Brennecke_getVariableGenes(Normalized_data$data)
```
**Figure 4**: 1,236 Significantly highly variable genes using all genes as spike-ins (10% FDR).

```{r}
heat_out = M3Drop_Expression_Heatmap(HVG, Normalized_data$data, 
		cell_labels = Normalized_data$labels)
```
**Figure 5***: Heatmap of expression of highly variable genes across cells.
