# Michaelis-Menten Modelling of Dropouts (M3D)

```{r}
library(M3D)
libary(M3DExampleData)
dim(Mmus_example_list$data)
length(Mmus_example_list$labels)
```

## Filtering & Normalizing Data

```{r}
norm_example_list = M3D_Clean_Data(Mmus_example_list$data, Mmus_example_list$labels, min_detected_genes=2000)
dim(norm_example_list$data)
```

## Fitting Michaelis-Menten Model
```{r fig.height=6 fig.width=6 fig.scap="Fitting MM, logistic, and double exponential models to dropouts."}
M3D_Dropout_Models(norm_example_list$data)
```

## Check For Biases
```{r fig.height=6 fig.width=6 fig.scap="Genes with CDS < 300 n.t."}
M3D_Test_Shift(norm_example_list$data, genes, name="Something")
```

## Identify Differentially Expressed Genes
```{r fig.height=6 fig.width=6 fig.scap="Significantly differentially expressed genes."}
DE_genes = M3D_Differential_Exptression(norm_example_list$data, mt_method="fdr", mt_threshold=0.01)
```

### Compare to Highly Variable Genes
```{r fig.height=6 fig.width=6 fig.scap="Highly variable genes using the Brennecke method."}
HVGs = Brennecke_getVariableGenes(norm_example_list$data, fdr=0.01)
data.frame(intersection= sum(HVGs %in% DE_genes[,1]), M3D = length(DE_genes[,1]), HVG= length(HVGs))
```

## Identify Cell Populations
```{r fig.height=7 fig.width=7 fig.scap="Gene expression profiles of DE genes."}
heat_output = M3D_Expession_Heatmap(DE_genes, norm_example_list$data, cell_labels = norm_example_list$labels)
clusters = M3D_Get_Heatmap_Cell_Clusters(heat_output, k=3)
```

### Marker Genes
```{r}
markers = M3D_getmarkers(norm_example_list$data, clusters)
markers[markers$AUC > 0.85,]
```
